packages/astro/src/default/utils/content.ts

```diff
@@ -75,250 +208,226 @@ export async function getTutorial(): Promise<Tutorial> {
-          );
-          continue;
-        }
-      }
-    }
+function assertTutorialStructure(tutorial: Tutorial) {
+  // verify that parts and lessons are not mixed in tutorial
+  if (Object.keys(tutorial.parts).length !== 0 && tutorial.lessons.some((lesson) => !lesson.part)) {
+    throw new Error(
+      'Cannot mix lessons and parts in a tutorial. Either remove the parts or move root level lessons into a part.',
+    );
   }
 
-  // sort lessons
-  lessons.sort((a, b) => {
-    const partsA = [
-      _tutorial.parts[a.part.id].order,
-      _tutorial.parts[a.part.id].chapters[a.chapter.id].order,
-      a.order,
-    ] as const;
-    const partsB = [
-      _tutorial.parts[b.part.id].order,
-      _tutorial.parts[b.part.id].chapters[b.chapter.id].order,
-      b.order,
-    ] as const;
-
-    for (let i = 0; i < partsA.length; i++) {
-      if (partsA[i] !== partsB[i]) {
-        return partsA[i] - partsB[i];
-      }
+  // verify that chapters and lessons are not mixed in a single part
+  for (const part of Object.values(tutorial.parts)) {
+    if (Object.keys(part.chapters).length === 0) {
+      continue;
     }
 
-    return 0;
-  });
-
-  const baseURL = import.meta.env.BASE_URL;
-
-  // now we link all lessons together
-  for (const [i, lesson] of lessons.entries()) {
-    const prevLesson = i > 0 ? lessons.at(i - 1) : undefined;
-    const nextLesson = lessons.at(i + 1);
-
-    const partMetadata = _tutorial.parts[lesson.part.id].data;
-    const chapterMetadata = _tutorial.parts[lesson.part.id].chapters[lesson.chapter.id].data;
+    if (tutorial.lessons.some((lesson) => lesson.part?.id === part.id && !lesson.chapter)) {
+      throw new Error(
+        `Cannot mix lessons and chapters in a part. Either remove the chapter from ${part.id} or move the lessons into a chapter.`,
+      );
+    }
+  }
+}
 
-    lesson.data = {
-      ...lesson.data,
-      ...squash(
-        [lesson.data, chapterMetadata, partMetadata, tutorialMetaData],
-        [
-          'mainCommand',
-          'prepareCommands',
-          'previews',
-          'autoReload',
-          'template',
-          'terminal',
-          'editor',
-          'focus',
-          'i18n',
-          'meta',
-          'editPageLink',
-          'openInStackBlitz',
-          'filesystem',
-        ],
-      ),
-    };
+function sortTutorialLessons(tutorial: Tutorial, metadata: TutorialSchema) {
+  const lessonIds = tutorial.lessons.map((lesson) => lesson.id);
 
-    if (prevLesson) {
-      const partSlug = _tutorial.parts[prevLesson.part.id].slug;
-      const chapterSlug = _tutorial.parts[prevLesson.part.id].chapters[prevLesson.chapter.id].slug;
+  // lesson ID alone does not make a lesson unique - combination of lessonId + chapterId + partId does
+  const lessonOrder: { lessonId: Lesson['id']; chapterId?: Chapter['id']; partId?: Part['id'] }[] = [];
```

Comment:
Argh it's frustrating that we can't use a `Map` in there. In theory we should but because in JS it only work with reference identity for objects (as we can't provide custom hash / equals functions), the only option would be to use a string that can be guaranteed to be unique or use an array like you did which doesn't suffer from the uniqueness issue.

At first I was going to suggest to use `:` to join the three things but then given that there won't be many, I'm more comfortable with a robust solution that uses an array.

Reply:
I think I had `-` separated IDs as key at one point here. 😄 
But in the last part where `tutorial.lessons` are looped, it was much nicer to directly match to objects instead of splitting the IDs again.

This function is quite complex but only because we provide so many ways to construct tutorials and customize the ordering. But I ended up building extensive set of test cases for all ordering cases I could think of:

https://github.com/stackblitz/tutorialkit/blob/0b2896a0e9e63a74ce8a3775523e9592e4b6b70f/packages/astro/src/default/utils/content.spec.ts#L284-L543